File: Widget/TodayViewController.m
//
//  TodayViewController.m
//  Widget
//
//  Created by Sachin on 7/30/14.
//  Copyright (c) 2014 Sachin Patel. All rights reserved.
//

#import "TodayViewController.h"

#import <CoreMotion/CoreMotion.h>
#import <NotificationCenter/NotificationCenter.h>
#import <QuartzCore/QuartzCore.h>

@interface TodayViewController () <NCWidgetProviding>

@property (nonatomic, weak) IBOutlet UILabel *stepCountLabel;
@property (nonatomic, weak) IBOutlet UILabel *mileCountLabel;
@property (nonatomic, weak) IBOutlet UILabel *floorCountLabel;

@property (nonatomic, weak) IBOutlet UIProgressView *todayProgressView;

@property (nonatomic) NSInteger stepCount;
@property (nonatomic) CGFloat mileCount;
@property (nonatomic) NSInteger floorCount;

@property (nonatomic, strong) NSMutableArray *lastWeek;
@property (nonatomic, strong) CMPedometer *pedometer;
@end

@implementation TodayViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setPreferredContentSize:CGSizeMake(320.0, 110.0f)];
    
    self.todayProgressView.layer.cornerRadius = 5.0;
    self.todayProgressView.layer.masksToBounds = YES;
    
    [[NSUserDefaults standardUserDefaults] synchronize];
    _stepCount = [[[NSUserDefaults standardUserDefaults] valueForKey:@"stepCount"] integerValue];
    [self updateLabels];
    
    if ([CMPedometer isStepCountingAvailable]) {
        _pedometer = [CMPedometer new];
        
        [_pedometer queryPedometerDataFromDate:[self midnightForDate:[NSDate date]] toDate:[NSDate date] withHandler:^(CMPedometerData *data, NSError *error) {
            _stepCount = data.numberOfSteps.integerValue;
            _mileCount = data.distance.floatValue * 0.000621371; // convert meters to miles
            _floorCount = data.floorsAscended.integerValue;
            
            [self updateLabels];
            
            [_pedometer startPedometerUpdatesFromDate:[NSDate date] withHandler:^(CMPedometerData *data, NSError *error) {
                _stepCount += data.numberOfSteps.integerValue;
                _mileCount += data.distance.floatValue * 0.000621371; // convert meters to miles
                _floorCount += data.floorsAscended.integerValue;
                
                [self updateLabels];
            }];
        }];
    }
}

- (NSDate *)midnightForDate:(NSDate *)date {
    NSCalendar *calendar = [NSCalendar autoupdatingCurrentCalendar];
    NSInteger units = (NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay);
    return [calendar dateFromComponents:[calendar components:units fromDate:date]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Core Motion

- (void)updateLabels {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            if (_stepCount < 10000) {
                _stepCountLabel.text = [NSString stringWithFormat:@"%ld", _stepCount];
            }
            else {
                NSNumber *number = [NSNumber numberWithDouble:_stepCount];
                NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
                [numberFormatter setNumberStyle:NSNumberFormatterDecimalStyle];
                [numberFormatter setGroupingSeparator:@","];
                _stepCountLabel.text = [numberFormatter stringForObjectValue:number];
            }
            
            _mileCountLabel.text = [NSString stringWithFormat:@"%.2f mile%@", _mileCount, _mileCount != 1.0 ? @"s" : @""];
            _floorCountLabel.text = [NSString stringWithFormat:@"%ld floor%@", _floorCount, _floorCount != 1 ? @"s" : @""];
            
            CGFloat percent = (CGFloat)_stepCount / 10000.0;
            _todayProgressView.progress = percent;
            
            if (percent < 0.60) {
                _todayProgressView.progressTintColor = [UIColor colorWithRed:255.0/255.0 green:59.0/255.0 blue:48.0/255.0 alpha:1.0];
            }
            else if (percent < 0.95) {
                _todayProgressView.progressTintColor = [UIColor colorWithRed:255.0/255.0 green:225.0/255.0 blue:0.0/255.0 alpha:1.0];
            }
            else {
                _todayProgressView.progressTintColor = [UIColor colorWithRed:76.0/255.0 green:217.0/255.0 blue:100.0/255.0 alpha:1.0];
            }
            
        });
    });
}

#pragma mark - Widget Handling

- (void)widgetPerformUpdateWithCompletionHandler:(void (^)(NCUpdateResult))completionHandler {
    [[NSUserDefaults standardUserDefaults] setValue:@(_stepCount) forKey:@"stepCount"];
    [[NSUserDefaults standardUserDefaults] synchronize];
    completionHandler(NCUpdateResultNewData);
}

- (UIEdgeInsets)widgetMarginInsetsForProposedMarginInsets:(UIEdgeInsets)defaultMarginInsets {
    return UIEdgeInsetsMake(0.0, 0.0, 0.0, 0.0);
}

@end

File: Widget/TodayViewController.swift
//
//  TodayViewController.swift
//  Steps
//
//  Created by Sachin Patel on 9/21/14.
//  Copyright (c) 2014 Sachin Patel. All rights reserved.
//

import Foundation
import NotificationCenter
import CoreMotion
import QuartzCore

class TodayViewController: UIViewController, NCWidgetProviding {
    
    // Interface
    let widgetHeight = 78.0
    @IBOutlet var stepCountLabel: UILabel!
    @IBOutlet var distanceLabel: UILabel!
    @IBOutlet var floorCountLabel: UILabel!
    @IBOutlet var progressView: UIProgressView!
    
    // Defines intervals for progress bar color
    let StepCountIntervalLow = Float(0)...Float(0.4)
    let StepCountIntervalMed = Float(0.4)...Float(0.8)
    
    // User Defaults
    var sharedDefaults: NSUserDefaults
    var unitSystemType: UnitSystem? {
        get {
            var raw = sharedDefaults.integerForKey(UnitTypeKey)
            return UnitSystem(rawValue: raw)
        }
        set {
            sharedDefaults.setInteger(newValue!.rawValue, forKey: UnitTypeKey)
            updateInterface()
        }
    }
    var unitDisplayType: UnitDisplay? {
        get {
            var raw = sharedDefaults.integerForKey(UnitDisplayKey)
            return UnitDisplay(rawValue: raw)
        }
        set {
            sharedDefaults.setInteger(newValue!.rawValue, forKey: UnitDisplayKey)
            updateInterface()
        }
    }
    var userGoal: Float {
        get {
            return sharedDefaults.floatForKey(UserGoalKey)
        }
        set {
            sharedDefaults.setFloat(newValue, forKey: UserGoalKey)
            updateInterface()
        }
    }
    var unitSystemWord: String {
        switch (unitSystemType!) {
        case .Imperial:
            switch (unitDisplayType!) {
                case .Short: return UnitSystemImperialWordShort
                case .Long: return UnitSystemImperialWord
            }
        case .Metric:
            switch (unitDisplayType!) {
            case .Short: return UnitSystemMetricWordShort
            case .Long: return UnitSystemMetricWord
            }
        }
    }
    
    // Pedometer
    var stepCount: Int
    var distance: Double
    var floorCount: Int
    var pedometer: CMPedometer
    
    // MARK: - Initializers
	
 	init() {
        stepCount = 0
        distance = 0.0
        floorCount = 0
        pedometer = CMPedometer()
        sharedDefaults = NSUserDefaults(suiteName: defaultsSuiteName)!
		super.init(nibName: nil, bundle: nil)
        preferredContentSize = CGSizeMake(UIScreen.mainScreen().bounds.width, CGFloat(widgetHeight))
    }
    
    required init(coder aDecoder: NSCoder) {
        stepCount = 0
        distance = 0.0
        floorCount = 0
        pedometer = CMPedometer()
        sharedDefaults = NSUserDefaults(suiteName: defaultsSuiteName)!
        super.init(coder: aDecoder)
        preferredContentSize = CGSizeMake(UIScreen.mainScreen().bounds.width, CGFloat(widgetHeight))
    }
    
    // MARK: - View lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set user defaults
        if userGoal == 0 {
            userGoal = 10_000
        }
        
        // Progress view appearance
        progressView.layer.cornerRadius = 8.0
        progressView.layer.masksToBounds = true
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        // Read from user defaults to show while updating data
        stepCount = sharedDefaults.integerForKey(StepCountKey)
        distance = sharedDefaults.doubleForKey(DistanceKey)
        floorCount = sharedDefaults.integerForKey(FloorCountKey)
        
        // Update interface before
        updateInterface()
        updatePedometer()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "updateInterface", name: NSUserDefaultsDidChangeNotification, object: nil)
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        
        sharedDefaults.setInteger(stepCount, forKey: StepCountKey)
        sharedDefaults.setDouble(distance, forKey: DistanceKey)
        sharedDefaults.setInteger(floorCount, forKey: FloorCountKey)
        sharedDefaults.synchronize()
        
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK: - Widget
    func updatePedometer () {
        // Ensure device has motion coprocessor
        if CMPedometer.isStepCountingAvailable() {
            
            // Get data from midnight to now
            pedometer.queryPedometerDataFromDate(NSDate.midnight, toDate: NSDate(), withHandler: {
                data, error in
                
                if data != nil {
                    // Store data
                    self.stepCount = data.numberOfSteps.integerValue
                    if CMPedometer.isDistanceAvailable() {
                        self.distance = data.distance.doubleValue
                    }
                    if CMPedometer.isFloorCountingAvailable() {
                        self.floorCount = data.floorsAscended.integerValue + data.floorsDescended.integerValue
                    }
                    self.updateInterface()
                }
                
                // Update as user moves
                self.pedometer.startPedometerUpdatesFromDate(NSDate(), withHandler: {
                    data, error in
                    
                    if data != nil {
                        // Add to existing counts
                        self.stepCount += data.numberOfSteps.integerValue
                        if CMPedometer.isDistanceAvailable() {
                            self.distance += data.distance.doubleValue
                        }
                        if CMPedometer.isFloorCountingAvailable() {
                            self.floorCount += data.floorsAscended.integerValue + data.floorsDescended.integerValue
                        }
                        self.updateInterface()
                    }
                })
            })
        }
    }
    
    func updateInterface () {
        dispatch_async(dispatch_get_main_queue(), {
            // Update step count label, format nicely
            if self.stepCount < 10_000 {
                self.stepCountLabel.text = "\(self.stepCount)"
            } else {
                var number = NSNumber(integer: self.stepCount)
                var formatter = NSNumberFormatter()
                formatter.numberStyle = .DecimalStyle
                formatter.groupingSeparator = ","
                self.stepCountLabel.text = formatter.stringForObjectValue(number)
            }
        });
        
        // Update distance label if available
        if CMPedometer.isDistanceAvailable() {
            // Update mile count label
            var convertedDistance = self.distance
            switch self.unitSystemType! {
            case .Imperial:
                convertedDistance /= mileInMeters
            case .Metric:
                convertedDistance /= 1000
            }
            
            var distanceExtension = (self.unitDisplayType == .Long && self.distance != 1) ? "s" : ""
            dispatch_async(dispatch_get_main_queue(), {
                self.distanceLabel.hidden = false
                self.distanceLabel.text = NSString(format: "%.2f %@%@", convertedDistance, self.unitSystemWord, distanceExtension) as String
            });
        } else {
            dispatch_async(dispatch_get_main_queue(), {
                self.distanceLabel.hidden = true
            });
        }
        
        // Update floor label if available
        if CMPedometer.isFloorCountingAvailable() {
            dispatch_async(dispatch_get_main_queue(), {
                self.floorCountLabel.hidden = false
                self.floorCountLabel.text = NSString(format: "%ld floor%@", self.floorCount, (self.floorCount != 1) ? "s" : "") as String
            });
        } else {
            dispatch_async(dispatch_get_main_queue(), {
                self.floorCountLabel.hidden = true
            });
        }
        
        // Update progress indicator
        var percent = min(Float(self.stepCount) / Float(self.userGoal), 1.0)
        self.progressView.setProgress(percent, animated: false)
        switch percent {
        case StepCountIntervalLow:
            self.progressView.progressTintColor = UIColor.redColor()
        case StepCountIntervalMed:
            self.progressView.progressTintColor = UIColor.yellowColor()
        default:
            self.progressView.progressTintColor = UIColor.greenColor()
        }
    }
    
    func widgetMarginInsetsForProposedMarginInsets(defaultMarginInsets: UIEdgeInsets) -> UIEdgeInsets {
        return UIEdgeInsetsMake(15.0, 47.0, 15.0, 15.0)
    }
}